### PR TITLE
Network Logger

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description
- implemented a logger in the `firmware/app` level thats inter-operable between sim/firmware
- created a network-logger that sends logs over the network
- updated `send_proto_over_udp.cpp` to show logs

### Testing Done
Unit tests + extensive manual testing + running sim

### Resolved Issues
#1652

### Length Justification

went _just_ over, so i think im in the clear here
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware_new/boards/frankie_v1/FreeRTOSConfig.h [hunk 1]
```diff
@@ -64,7 +64,7 @@ void xPortSysTickHandler(void);
 #define configTICK_RATE_HZ ((TickType_t)1000)
 #define configMAX_PRIORITIES (56)
 #define configMINIMAL_STACK_SIZE ((uint16_t)1024)
-#define configTOTAL_HEAP_SIZE ((size_t)30720)
+#define configTOTAL_HEAP_SIZE ((size_t)80720)
```
Comment (jonathanlew) [inline src/firmware_new/boards/frankie_v1/FreeRTOSConfig.h:67]: why this size in particular?

Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/FreeRTOSConfig.h:67]: I just "downloaded" 50kb more ram, nothing more
once we finalize a bit more things, we can justify with https://github.com/UBC-Thunderbots/Software/issues/1192

File: src/firmware/app/world/firmware_robot.c [hunk 2]
```diff
@@ -52,6 +54,8 @@ FirmwareRobot_t* app_firmware_robot_create(
     new_robot->robot_constants            = robot_constants;
     new_robot->controller_state           = controller_state;
 
+    TLOG_INFO("new robot has been created!");
```
Comment (akhilveeraghanta) [inline src/firmware/app/world/firmware_robot.c]: this is for demo purposes, this log will show up on sim, will be removed before merging this pr

File: src/firmware/app/logger/logger.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * Initializes the logger with the given robot id and the log handler function
+ *
+ * @param robot_id The robot id the logs are coming from
+ * @param robot_log_msg_handler The function to handle RobotLog protos
+ */
+void app_logger_init(unsigned robot_id,
+                     void (*robot_log_msg_handler)(TbotsProto_RobotLog log_msg));
+
+/**
+ * Creates a new RobotLog proto and calls the the robot_log_msg_handler
```
Comment (Resocram) [inline src/firmware/app/logger/logger.h]: Could you explain how the logging process will work? Do we call app_logger_init for each robot and if we want to log something for that particular robot, we call either TLOG_DEBUG, TLOG_INFO, etc.?
Why do lines 27-34, have the hard-coded file and line values, is every log that is going to logged in the same file and line number? 
 (Also there are two "the"s in line 14)

Comment (akhilveeraghanta) [reply src/firmware/app/logger/logger.h]: > Could you explain how the logging process will work? Do we call app_logger_init for each robot and if we want to log something for that particular robot, we call either TLOG_DEBUG, TLOG_INFO, etc.?

Yup thats it, `app_logger_init` will already be called in the main function before any tasks start running, so we just use the macros to log

> Why do lines 27-34, have the hard-coded file and line values, is every log that is going to logged in the same file and line number?

`__LINE__` and `__FILE__` are predefined macros that the pre-processor will replace with the line number and file name. So the log will contain info about the line number and file name of where its being used. 
Added a comment to clarify

>  (Also there are two "the"s in line 14)
Fixed!

File: src/firmware_new/boards/frankie_v1/io/network_logger.c [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+{
+    // holds the msg that was just dequeued
+    TbotsProto_RobotLog dequeued_robot_log_proto;
+
+    ProtoMulticastCommunicationProfile_t* profile =
+        (ProtoMulticastCommunicationProfile_t*)communication_profile;
+
+    osStatus_t status;
+
+    for (;;)
```
Comment (Resocram) [inline src/firmware_new/boards/frankie_v1/io/network_logger.c:27]: What is the purpose of this io_network class? And is it okay to leave this running in an infinite loop?

Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/io/network_logger.c:27]: the io_network logger is responsible for consuming the queue of robot log msgs and sending them over the network

the infinite loop is a fairly common pattern in FreeRTOS, you want this task to keep doing this as long as the robot is alive. The `osMessageQueueGet` function blocks until there is a msg so it doesn't actually use up _too_ much resources

**Comment by Resocram** [comment]
The Travis build is failing in one of the tests, memory leak check, and formatting issue, is it possible to fix them? Otherwise LGTM!

File: src/shared/proto/robot_log_msg.proto [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+import "shared/proto/tbots_timestamp_msg.proto";
+import "nanopb/options.proto";
+
+// the severity level of the log
+enum LogLevel
+{
+    DEBUG = 0;
+    INFO  = 1;
+    WARN  = 2;
```
Comment (jonathanlew) [inline src/shared/proto/robot_log_msg.proto]: to be consistent with g3log:
```suggestion
    WARNING  = 2;
```

**Review by jonathanlew — COMMENTED** [review]
Overall looks good. It would be good to discuss how the the robot logs should be displayed, but that might be out of scope. Please fix the TODOs attached to #1652

File: src/shared/proto/robot_log_msg.proto (hunk 5 above)
Comment (akhilveeraghanta) [reply src/shared/proto/robot_log_msg.proto]: fixed

**Review by Resocram — DISMISSED** [review]

File: src/software/simulation/simulator_robot_singleton_test.cpp [hunk 6]
```diff
@@ -21,6 +24,11 @@ extern "C"
 class SimulatorRobotSingletonTest : public testing::Test
 {
    protected:
+    void SetUp() override
+    {
+        app_logger_init(0, &TestUtil::handleTestRobotLog);
```
Comment (jonathanlew) [inline src/software/simulation/simulator_robot_singleton_test.cpp:32]: `SimulatorRobotSingleton` has a `handleYellowRobotLogProto`, why can't we use that?

**Review by jonathanlew — COMMENTED** [review]
One last comment here: we have 3 instances of `LOG(INFO)` that are very similar. Can we make a common file and define the things they have in common?

File: src/software/simulation/simulator_robot_singleton_test.cpp (hunk 6 above)
Comment (jonathanlew) [reply src/software/simulation/simulator_robot_singleton_test.cpp:32]: A bit more documentation would help

Comment (akhilveeraghanta) [reply src/software/simulation/simulator_robot_singleton_test.cpp:32]: added comment here, and created a ticket to remove the duplication

File: src/firmware/app/logger/logger.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                     void (*robot_log_msg_handler)(TbotsProto_RobotLog log_msg));
+
+/**
+ * Creates a new RobotLog proto and calls the robot_log_msg_handler
+ *
+ * NOTE: If the log msg is longer than the max_size defined in the proto
+ * the log will be truncated.
+ *
+ * @param file_name The file this log is located
+ * @param line_number The line number where this log is located
```
Comment (EvanMorcom) [inline src/firmware/app/logger/logger.h]: Nit:

Kinda felt like the 'line number where the log is located` sounds a bit clunky.

Maybe 'line where the log will be printed?', also what if there is a newline character, is it possible to be on multiple lines.

File: src/firmware/app/logger/logger.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * @param log_level The severity level of this log
+ * @param format The format string followed by printf style args
+ */
+void app_logger_log(const char *file_name, unsigned line_number,
+                    TbotsProto_LogLevel log_level, const char *format, ...);
+
+
+// `__LINE__` and `__FILE__` are predefined macros that the preprocessor will replace
```
Comment (EvanMorcom) [inline src/firmware/app/logger/logger.h:28]: Do these macros use the `file_name` parameter?

**Review by jonathanlew — DISMISSED** [review]
LGTM other than other reviewers' comments

File: src/firmware/app/logger/logger.h (hunk 8 above)
Comment (akhilveeraghanta) [reply src/firmware/app/logger/logger.h:28]: yup they do, the file_name is set to `__FILE__`

File: src/firmware/app/logger/logger.h (hunk 7 above)
Comment (akhilveeraghanta) [reply src/firmware/app/logger/logger.h:20]: fixed!

Comment (akhilveeraghanta) [reply src/firmware/app/logger/logger.h]: fixed

**Review by jonathanlew — APPROVED** [review]
= 🪓 + 🌳

**Review by Resocram — APPROVED** [review]
LGTM!

**Pull request merged**